### ISSUE TITLE
Store in-flight ship/outfit attributes separately from for-construction attributes

### ISSUE THREAD

**Issue by petervdmeer** [endless-sky/endless-sky#8710]
### Problem Description

String comparisons (in the ships and some outfit attributes) slow down the game when the player is in-flight.

[PR 7685; Hash-based attributes](https://github.com/endless-sky/endless-sky/pull/7685) tries to address this slow-down by using faster hash-based comparisons.
[PR 7165; Vector-indexed attributes](https://github.com/endless-sky/endless-sky/pull/7685) tries to address this slow-down by using vector(integer)-based indexing.

Both PRs contain very good ideas to improve performance, but both also get some negative effects from the potentially large set of attributes that we can have, because [any outfit can define new custom attributes](based on https://github.com/endless-sky/endless-sky/blob/a81b103007baed68f1d3087784b42cacf6163b0f/source/Outfit.cpp#L297-L298). This also means that the full list of attributes is not known at compile time.

### Related Issue Links

- [6329 Performance Bottlenecks](https://github.com/endless-sky/endless-sky/issues/6329) triggered many relevant discussions because it mentions 11.5% of the time being spent on String comparisons.
- [5784; separate string interner](https://github.com/endless-sky/endless-sky/pull/5748) is somewhat related, because it would replace string comparisons with char* comparisons. (Which might be faster, but doesn't address the potentially large list lookups.)
- The PRs mentioned in the problem description are obviously related.
- [Tehhowch comment](https://github.com/endless-sky/endless-sky/pull/6204#pullrequestreview-735453664) also seems somewhat related.

### Desired Solution

My suggestion is to split the attributes into "in-flight" and "at compose" time. ("at compose" is when composing ships at the outfitter, or when plundering (decomposing) hostile ships) The list of attributes that are used "in-flight" should be limited and known at compile time, and get the performance improvements that PRs 7685 and 7165 tried to achieve. (Probably vector/integer-based indexing for ship attributes, not sure yet what to do for "ammo outfits", maybe map<int, value> since they are used when using ammo when firing a weapon.)

### Alternative Approaches

Improvements to the attributes as-is (as done in the mentioned PRs) are an alternative.

### Additional Context

A complexity is that the game currently supports using any Outfit as ammo for another weapon-outfit; we can theoretically have weapons that launch "Fission Reactors" or weapons that launch "Quantum Keystones" or even weapons that launch "Torpedo Launchers".
[Ammo usage is currently dealt with as a full outfit removal](https://github.com/endless-sky/endless-sky/blob/a81b103007baed68f1d3087784b42cacf6163b0f/source/Ship.cpp#L4077), but when we launch a weapon during flight, then only the performance critical attributes need to be re-calculated. (The attributes not used in flight can be re-calculated at landing.)

**Comment by Amazinite** [comment]
Is something like #6804 what we're looking for? It doesn't cover everything, but it gets most of the existing calls to a ship's attributes that are done every frame. One of the big remaining sticking points I mentioned in the comments was with jump-related attributes, but that's since been covered by #7378, so if we did something like #6804, the only remaining attribute gets that happen each frame would be required crew and drag, at least according to one of my comments on that PR from May of last year.

**Comment by petervdmeer** [comment]
> Is something like #6804 what we're looking for?

Putting the EnergyLevels in their own classes sounds like a good first step in the right direction. We will probably need a few more PRs on top of those changes (and maybe remove the classes at some point), but it is a good start.

**Comment by tibetiroka** [comment]
Related: #9025. It would allow using simple map operations instead of multiple individual lookups, and completely avoids string comparisons for most attributes.

**Comment by petervdmeer** [comment]
> Related: #9025. It would allow using simple map operations instead of multiple individual lookups, and completely avoids string comparisons for most attributes.

Agreed, using such a structure (regardless of if we use it for the keywords as well) can also help.
